horovod/run/elastic/discovery.py

```diff
@@ -0,0 +1,139 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import logging
+import threading
+
+from collections import defaultdict
+
+import six
+
+from horovod.run.common.util import safe_shell_exec
+
+
+class HostState(object):
+    def __init__(self):
+        self._event = threading.Event()
+
+        # TODO(travis): blacklisted hosts should have a timeout period that increases with each failure
+        self._blacklisted = False
+
+    def get_event(self):
+        if self._event.is_set():
+            event = threading.Event()
+            self._event = event
+        return self._event
+
+    def set_event(self):
+        self._event.set()
+
+    def blacklist(self):
+        self._blacklisted = True
+        self.set_event()
+
+    def is_blacklisted(self):
+        return self._blacklisted
+
+
+class DiscoveredHosts(object):
+    def __init__(self, discovery):
+        self._available_hosts = set()
+        self._available_slots = {}
+        self._host_state = defaultdict(HostState)
+        self._discovery = discovery
+
+    def update_available_hosts(self):
+        prev_hosts = self._available_hosts
+        prev_slots = self._available_slots
+        available_hosts, available_slots = self._discovery.find_available_hosts_and_slots()
+        if prev_hosts != available_hosts or prev_slots != available_slots:
+            self._available_hosts, self._available_slots = available_hosts, available_slots
+            return True
+        return False
+
+    def count_available_slots(self):
+        return sum([self.get_slots(host) for host in self._available_hosts
+                    if not self._host_state[host].is_blacklisted()])
+
+    def get_slots(self, host):
+        return self._available_slots[host]
+
+    def filter_available_hosts(self, hosts):
+        return [host for host in hosts
+                if host in self._available_hosts and not self._host_state[host].is_blacklisted()]
+
+    def get_available_hosts(self):
+        return self.filter_available_hosts(self._available_hosts)
+
+    def blacklist(self, host):
+        if not self._host_state[host].is_blacklisted():
+            logging.warning('blacklist failing host: {}'.format(host))
+        self._host_state[host].blacklist()
+
+    def is_blacklisted(self, host):
+        return self._host_state[host].is_blacklisted()
+
+    def count_blacklisted_slots(self):
+        return sum([self.get_slots(host) for host, meta in self._host_state.items() if meta.is_blacklisted()])
+
+    def get_host_event(self, host):
+        return self._host_state[host].get_event()
+
+
+class HostDiscovery(object):
+    def find_available_hosts_and_slots(self):
```

Comment:
Isn't `hosts = set(slots.keys())` always the case? would it be sufficient to return slots only then?

Reply:
Yes, this was a holdover from an earlier design.  Refactored and removed the `hosts` variables.

Reply:
please document the contract: now it is a dict(hostname -> number of slots)